Detectors/CTP/simulation/src/Digits2Raw.cxx

```diff
@@ -0,0 +1,181 @@
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+/// \file Digits2Raw.cxx
+/// \author Roman Lietava
+
+#include "CTPSimulation/Digits2Raw.h"
+#include "FairLogger.h"
+#include "CommonUtils/StringUtils.h"
+
+using namespace o2::ctp;
+
+void Digits2Raw::init()
+{
+  // CTP Configuration
+  if (mCCDBServer.empty()) {
+    LOG(FATAL) << "CTP digitizer: CCDB server is not set";
+  } else {
+    LOG(INFO) << "CTP digitizer:: CCDB server:" << mCCDBServer;
+  }
+  auto& mgr = o2::ccdb::BasicCCDBManager::instance();
+  mgr.setURL(mCCDBServer);
+  mCTPConfiguration = mgr.get<CTPConfiguration>(o2::ctp::CCDBPathCTPConfig);
+  //
+  // Register links
+  //
+  std::string outd = mOutDir;
+  if (outd.back() != '/') {
+    outd += '/';
+  }
+  for (int ilink = 0; ilink < mNLinks; ilink++) {
+    mFeeID = uint64_t(ilink);
+    std::string outFileLink = mOutputPerLink ? o2::utils::Str::concat_string(outd, "ctp_link", std::to_string(ilink), ".raw") : o2::utils::Str::concat_string(outd, "ctp.raw");
+    mWriter.registerLink(mFeeID, mCruID, ilink, mEndPointID, outFileLink);
+  }
+  mWriter.setEmptyPageCallBack(this);
+}
+void Digits2Raw::processDigits(const std::string& fileDigitsName)
+{
+  std::unique_ptr<TFile> digiFile(TFile::Open(fileDigitsName.c_str()));
+  if (!digiFile || digiFile->IsZombie()) {
+    LOG(FATAL) << "Failed to open input digits file " << fileDigitsName;
+    return;
+  }
+  TTree* digiTree = (TTree*)digiFile->Get("o2sim");
+  if (!digiTree) {
+    LOG(FATAL) << "Failed to get digits tree";
+    return;
+  }
+  std::vector<o2::ctp::CTPDigit> CTPDigits, *fCTPDigitsPtr = &CTPDigits;
+  if (digiTree->GetBranch("CTPDigits")) {
+    digiTree->SetBranchAddress("CTPDigits", &fCTPDigitsPtr);
+  } else {
+    LOG(FATAL) << "Branch CTPDigits is missing";
+    return;
+  }
+  // Get first orbit
+  digiTree->GetEntry(0);
+  uint32_t orbit0;
+  if (CTPDigits.size() > 0) {
+    orbit0 = CTPDigits[0].intRecord.orbit;
+  } else {
+    LOG(FATAL) << "Branch CTPDigits there but no entries";
+    return;
+  }
+  // Add all CTPdigits for given orbit
+  for (int ient = 1; ient < digiTree->GetEntries(); ient++) {
+    digiTree->GetEntry(ient);
+    int nbc = CTPDigits.size();
+    LOG(INFO) << "Entry " << ient << " : " << nbc << " BCs stored";
+    std::vector<std::bitset<NGBT>> hbf;
+    for (auto const& ctpdig : CTPDigits) {
+      if (orbit0 == ctpdig.intRecord.orbit) {
+        std::bitset<NGBT> gbtdig;
+        digit2GBTdigit(gbtdig, ctpdig);
+        hbf.push_back(gbtdig);
+      } else {
+        std::vector<char> buffer;
+        buffer = digits2HBTPayload(hbf, NIntRecPayload);
+        std::string PLTrailer;
+        std::memcpy(buffer.data() + buffer.size() - o2::raw::RDHUtils::GBTWord, PLTrailer.c_str(), o2::raw::RDHUtils::GBTWord);
+        orbit0 = ctpdig.intRecord.orbit;
+      }
```

Comment:
I don't see where do you call `writer.addData(...)`.
BTW, why do you watch orbit change? You just need to call addData for every orbit/bc where you have non-0 payload (and if you call it for 1 link, then this should be done for all. There is a writer-side protection against not doing this, but you better take care of this on your side).

Reply:
> Orbit: for non zero suppressed data I need to add (bcid, 0) for all bcs which are not in input record.

OK, but it should be done in a single loop, w/o pre-reading the entry. 

> Question on emptyHBFMethod: how can I do it link dependent ?

You receive in the emptyHBFMethod the rdh for which the payload need to be generated, so you have the full info about the link.

Reply:
@shahor02 what is PLTrailer ? Do I need it ?